### PR TITLE
Audio in, memorypcm, noclone merged for AMYchip support

### PR DESCRIPTION
**Pull request by bwhitman**
This merges together #230, #229, #228 so we can work on one branch for [amychip](https://github.com/shorepine/amychip) dev without conflicts.

- remove clone_osc
- memorypcm inside AMY
- audio in support, use `wave=AUDIO_IN0` or `AUDIO_IN1`. 
- wire sample transfer support, `z` / `load_sample` using base64 encoding
- docs

### CODE REVIEW

**Comment by bwhitman** [comment]
Some feedback from dan we'll implement
 - pcm patch numbers for memorypcm can be anything, not 1024+, if set to an existing flash patch will just play the memorypcm patch instead
 - we need to unload all samples on reset(RESET_ALL_OSCS)
 - we need to unload a sample on reset(osc) 
 - audio in should be "on the virtual mixer/bus plane" see #114

**Comment by bwhitman** [comment]
@dpwe I wonder if we should merge this soon, just the audioin and memorypcm stuff, then think about buses and sequencer in a separate PR? It's getting big. The audioin stuff can stay "hidden" (supported but no API) until we do the buses/better API for it.

File: src/amy.h [hunk 1]
```diff
@@ -98,7 +102,8 @@ enum coefs{
 #define AUDIBLE_SUSPENDED 4
 #define IS_MOD_SOURCE 5
 #define IS_ALGO_SOURCE 6
-#define STATUS_OFF 7
+#define TRANSFER_DATA 7
```
Comment (dpwe) [inline src/amy.h:105]: We have weirdly mixed together osc[].status values with event.status.  I don't think there's any reason to mix them - I think event.status is only EMPTY, SCHEDULED, and now TRANSFER_DATA, and synth.status is the rest?

**Review by dpwe — APPROVED** [review]
Looks great.  I didn't look too closely at pcm.c, transfer.c, or miniaudio, but I'm sure it's all good

File: src/amy.c [hunk 2]
```diff
@@ -580,65 +580,15 @@ void amy_add_event_internal(struct event e, uint16_t base_osc) {
...
-    //synth[i].last_two[1] = synth[f].last_two[1];
-    //synth[i].lut = synth[f].lut;
-}
+
 
 void reset_osc(uint16_t i ) {
+    // reset PCM patch if type = PCM
+    if(synth[i].wave == PCM) {
+        if(AMY_IS_SET(synth[i].patch)) {
+            pcm_unload_patch(synth[i].patch);
```
[the remark this answers is not in the document]
Comment (bwhitman) [reply src/amy.c:589]: OK, we now unalloc all samples on amy.reset(RESET_ALL_OSCS)  and unalloc a single patch with amy.unload_sample(patch)

File: src/amy.h (hunk 1 above)
Comment (bwhitman) [reply src/amy.h:105]: This is fixed up now

File: src/amy.c [hunk 3]
```diff
@@ -884,6 +838,20 @@ void oscs_deinit() {
...
 }
 
+void audio_in_note_on(uint16_t osc, uint8_t channel) {
+    // do i need to do anything here? probably not
+}
+
+SAMPLE render_audio_in(SAMPLE * buf, uint16_t osc, uint8_t channel) {
+    uint16_t c = 0;
+    for(uint16_t i=channel;i<AMY_BLOCK_SIZE*AMY_NCHANS;i=i+AMY_NCHANS) {
+        buf[c] = F2S(((float)amy_in_block[i]/32767.0) * msynth[osc].amp);
```
[the remark this answers is not in the document]
Comment (bwhitman) [reply src/amy.c]: That works great

**Comment by bwhitman** [comment]
thank you dan!! gonna merge this and start a new branches for sequencer (and later bus)

**Pull request merged**